### ISSUE TITLE
The compiler should error when trying to transmute pointer values with incompatible alignment

### ISSUE THREAD

**Issue by hkBst** [rust-lang/rust#95479]
I tried this code:

```rust
struct Foo {
    a: i32,
    b: i32,
}

fn main() {
    unsafe {
        let mut array: [u8; 1024] = [1; 1024];
        let foo = std::mem::transmute::<&mut u8, &mut Foo>(&mut array[0]);
        foo.a += 1;
    }
}
```

I expected to see this happen: the compiler complains about the alignment of the array not being good enough for the struct Foo

Instead, this happened: the compiler did not complain

### Meta

`rustc --version --verbose`:
```
1.59.0 from the playground
```

This issue was noticed by the creator of Zig in this blog post: https://andrewkelley.me/post/unsafe-zig-safer-than-unsafe-rust.html.

**Comment by leonardo-m** [comment]
Yes, here the Zig type system has static guarantee stronger than Rust ones because it (optionally?) keeps track of alignments.
